lib/vquic/ngtcp2.c

```diff
@@ -972,15 +1094,24 @@ static ssize_t ngh3_stream_recv(struct connectdata *conn,
     return -1;
   }
 
-  if(stream->memlen) {
+  if(stream->memlen || stream->overflow_buflen) {
+    /* if there's data in the overflow buffer and room in the receive buffer,
+       copy as much as possible to the receive buffer before returning */
+    drain_overflow_buffer(stream);
     ssize_t memlen = stream->memlen;
     /* data arrived */
     *curlcode = CURLE_OK;
     /* reset to allow more data to come */
     stream->memlen = 0;
     stream->mem = buf;
     stream->len = buffersize;
+    /* extend the stream window with the data we're consuming and send out
+       any additional packets to tell the server that we can receive more */
     extend_stream_window(qs->qconn, stream);
+    if(ng_flush_egress(conn, sockfd, qs)) {
```

Comment:
This was missing from the change in #4600. This was causing the `MAX_STREAM_DATA` frames that `extend_stream_window` was generating to not be sent right away. If the server had already reached the flow control limit we would never receive another packet from it and we'd never actually send these frames with the ng_flush_egress call above, so the transfer would hang.